streaming/receiver.c

```diff
@@ -0,0 +1,425 @@
+#endif
+        log_stream_connection(rpt->client_ip, rpt->client_port, rpt->key, rpt->host->machine_guid, rpt->host->hostname, "FAILED - CANNOT REPLY");
+        error("STREAM %s [receive from [%s]:%s]: cannot send ready command.", rpt->host->hostname, rpt->client_ip, rpt->client_port);
+        close(rpt->fd);
+        return 0;
+    }
+
+    // remove the non-blocking flag from the socket
+    if(sock_delnonblock(rpt->fd) < 0)
+        error("STREAM %s [receive from [%s]:%s]: cannot remove the non-blocking flag from socket %d", rpt->host->hostname, rpt->client_ip, rpt->client_port, rpt->fd);
+
+    // convert the socket to a FILE *
+    FILE *fp = fdopen(rpt->fd, "r");
+    if(!fp) {
+        log_stream_connection(rpt->client_ip, rpt->client_port, rpt->key, rpt->host->machine_guid, rpt->host->hostname, "FAILED - SOCKET ERROR");
+        error("STREAM %s [receive from [%s]:%s]: failed to get a FILE for FD %d.", rpt->host->hostname, rpt->client_ip, rpt->client_port, rpt->fd);
+        close(rpt->fd);
+        return 0;
+    }
+
+    rrdhost_wrlock(rpt->host);
+/* if(rpt->host->connected_senders > 0) {
+        rrdhost_unlock(rpt->host);
+        log_stream_connection(rpt->client_ip, rpt->client_port, rpt->key, rpt->host->machine_guid, rpt->host->hostname, "REJECTED - ALREADY CONNECTED");
+        info("STREAM %s [receive from [%s]:%s]: multiple streaming connections for the same host detected. Rejecting new connection.", rpt->host->hostname, rpt->client_ip, rpt->client_port);
+        fclose(fp);
+        return 0;
+    }
+*/
+
+    rrdhost_flag_clear(rpt->host, RRDHOST_FLAG_ORPHAN);
+//    rpt->host->connected_senders++;
+    rpt->host->senders_disconnected_time = 0;
+    rpt->host->labels_flag = (rpt->stream_version > 0)?LABEL_FLAG_UPDATE_STREAM:LABEL_FLAG_STOP_STREAM;
+
+    if(health_enabled != CONFIG_BOOLEAN_NO) {
+        if(alarms_delay > 0) {
+            rpt->host->health_delay_up_to = now_realtime_sec() + alarms_delay;
+            info("Postponing health checks for %ld seconds, on host '%s', because it was just connected."
+            , alarms_delay
+            , rpt->host->hostname
+            );
+        }
+    }
+    rrdhost_unlock(rpt->host);
+
+    // call the plugins.d processor to receive the metrics
+    info("STREAM %s [receive from [%s]:%s]: receiving metrics...", rpt->host->hostname, rpt->client_ip, rpt->client_port);
+    log_stream_connection(rpt->client_ip, rpt->client_port, rpt->key, rpt->host->machine_guid, rpt->host->hostname, "CONNECTED");
+
+    cd.version = rpt->stream_version;
+
+
+    size_t count = streaming_parser(rpt, &cd, fp);
+    //size_t count = pluginsd_process(host, &cd, fp, 1);
+
+    log_stream_connection(rpt->client_ip, rpt->client_port, rpt->key, rpt->host->machine_guid, rpt->host->hostname, "DISCONNECTED");
+    error("STREAM %s [receive from [%s]:%s]: disconnected (completed %zu updates).", rpt->host->hostname, rpt->client_ip, rpt->client_port, count);
```

Comment:
The fix now now timeouts the old connection in 30 seconds and accepts the new one. But the old one doesn't go away. It is marked as being shutdown, but the thread is there, stuck in fgets

```
#1  __GI___libc_read (fd=213, buf=0x561448eee0d0, nbytes=4096) at ../sysdeps/unix/sysv/linux/read.c:24
24	in ../sysdeps/unix/sysv/linux/read.c

 :
 :
#8  0x0000561444e85478 in streaming_parser (rpt=0x561448ca5060, cd=0x7fe5692c7810, fp=0x561448eec1a0) at streaming/receiver.c:170
170	        if (receiver_read(rpt, fp))
(gdb) p rpt->shutdown 
$1 = 1
```

If for whatever reason the fgets returns, it will immediately disconnect but it shouldn't execute the code in L393-L401 (maybe the `rpt->host->receiver  == rpt` test is enough to protect that part)

Reply:
This is what can actually crash the agent after the socket times out. Maybe there is a way to force the socket to die somehow. Maybe try [shutdown](https://www.man7.org/linux/man-pages/man2/shutdown.2.html)?

Reply:
When the `fgets` retiurns it will not touch the host structure, only the `receiver_state` which has been detached from the host. I've added the `shudown()` call to help along towards its destination.